benchadapt/python/benchadapt/machine_info.py

```diff
@@ -73,9 +74,14 @@ def python_info():
 def github_info():
     commit = _exec_command(["git", "rev-parse", "HEAD"])
     repository = _exec_command(["git", "remote", "get-url", "origin"])
+    branch = _exec_command(["git", "branch", "--show-current"])
+    remote = _exec_command(["git", "remote", "get-url", "origin"])
+    fork = re.search("(?<=.com[/:])([^/]*?)(?=/)", remote).group(0)
+
     return {
         "commit": commit,
         "repository": repository.rsplit(".git")[0],
+        "branch": f"{fork}:{branch}",
```

Comment:
oh right I got it backwards - what I meant is that ideally this logic only lives in the adapters package. We'll be removing `conbench/machine_info.py` to the adapters package eventually. But this PR touches both `conbench/machine_info.py` and `benchadapt/python/benchadapt/machine_info.py`

Reply:
> ideally this logic only lives in the adapter

Is that the ideal state that we want? I definitely agree that having it in one place and not two is good — but I'm not totally sure that the adapter is the right place for this compared to the backend. 

One nice thing about keeping this in the backend is that we have a bit more control over the details of the implementation (and since our comparison flow is also in the backend we can keep them more closely aligned — or change them without needing to go through both the adapter and the backend). It also keeps this logic slightly closer (at least in the same component) to the places that will subsequently use that information (e.g. comparisons). 

What are the pros/cons of putting this in the adapter versus the backend? We also can punt on the specifics decision for now and document some of these questions in an issue and keep the code we have now in both places

Reply:
By "backend" you mean the Conbench server, right? Unfortunately I think it's impossible to figure out the GitHub repo name, fork name, branch name, and commit SHA from the server side. Even if the client provided the repo and SHA, it would be fairly expensive or impossible in some cases to figure out the fork and branch using the GitHub API. The client has cheap access to all this information, which is why we've always been figuring it out on the client side.

Reply:
> By "backend" you mean the Conbench server, right?

Yeah, exactly.

But if the client provides the pull number all of that info is readily available from the pulls endpoint, no? For example: https://api.github.com/repos/apache/arrow/pulls/14560 has `head.label` for fork+branch name (or `head.repo.full_name` + `head.ref` for them separately), as well as the fork_point_sha in `base.sha`. Is that endpoint missing anything that we need?

Since these runs are fundamentally orchestrated from pullrequests, having the PR number should be trivial during the orchestration to run them (I assume we'll still need to send the SHA there too...). And all of the feedback about the outcome of the run will be PR-based anyway with benchalerts, so we'll need the PR number there anyway.

We might _some day_ need to support random-branch benchmarks that aren't associated with PRs, but we absolutely do not need to support them today. And if that lets us avoid things like figuring out how to reliably parse remotes in CI or whatever, IMO that's a win.

Reply:
I thought we had agreed to go with branch because PR# would have to be manually specified by the user and inserted (i.e. you've got to munge results a lot), whereas branch can be detected

Reply:
I don't remember that we decided on that level of detail — but do remember distinctly talking about how "PR# would have to be manually specified by the user" isn't quite right for our current use case, like I said above: the current use should always have a PR number easily accessible, and if that makes all the info easy to get from that endpoint we should take advantage of that.

Reply:
But, regardless of if we did or didn't decide to use one end point or the other: neither of those are themselves requirements — what we care about is the external behavior and then what it takes code-wise to produce that behavior. And my point in this comment is: let's weight the pros and cons of each of these end points and decide which one will make a better interface. Even if we decided in the past to use one, that might not make it the right one now or in the future.

Reply:
My interest here is to minimize configuration. PR# is known in GHA, so it's another piece of metadata like `run_reason` that needs to be passed through to the results in every implementation. Branch we can automate once for everyone and be done with, and they can override it if they want.

If we care about the PR# we can do it; I'd just rather save some work if we end up with the same thing.

Reply:
We also don't have an absolute guarantee that when benchmarks are being run, they have a git checkout that can have `git` operations made against (I know of at least one conbench setup that uses a pre-built binary in the benchmark phase, so doesn't have the git repo available). But IIUC benchalerts needs the PR number to operate anyway, so that seems like the lowest-common denominator (additionally, describing the requirement of what to seen is pretty easy to communicate to others implementing this stuff "you need the github pullrequest number" versus "you need the upstream branch + fork formatted in this specific way, you also need to get the fork-merge-sha which is...")

Reply:
We also don't need to decide this here and now – this PR can go with both of these in place until we decide on which location really is the right on for getting this information in the modal case.

Reply:
Yeah, if it could be an either `branch` or `pr` needs to be specified (along the lines of `stats`/`error` and `machine_info`/`cluster_info`, that's viable. It's a tad more work to handle and validate, but since we've already got similar logic in the schema it's probably alright

Reply:
> Yeah, if it could be an either branch or pr needs to be specified (along the lines of stats/error and machine_info/cluster_info, that's viable

Yup, also an option, and that has the benefit of letting us see how they work + which is easier to work with (with the possible downside of being more complex like you mention or having someone somewhere depend on one that we want to deprecate later, but in the abstract at least, those two aren't so bad — we might get into the details and realize it's waaaay more work to maintain both, but I don't think it will be from here)

Reply:
You two are commenting too fast for me to keep up :)

So it sounds like with either option, the client must scrape the repo name and commit sha from a local cloned repo. In order to get the repo name, we have to figure out which remote the current branch is tracking (currently on main this is [hardcoded](https://github.com/conbench/conbench/blob/564bc7c0498e06a4ca260e8ce050304a506c06c7/conbench/machine_info.py#L83) to `origin`) and get that remote's github URL. This is doable without assuming anything about the git setup. (Though, note that we've never run into problems with `origin` before.)

If there isn't a local cloned repo, it's up to the runner to supply those two values manually.

The two options define how we get the fork name and branch name. Option 1 is the client collects the fork and branch from the cloned repo in a similar way to collecting the repo and sha. Option 2 is the client provides the PR number and the server figures out the fork and branch from hitting the API.

Pros of 1: in the above process of collecting the repo name and commit sha, we already have the branch and fork handy.
Cons of 1: if there isn't a local cloned repo, the runner configurer must know how to supply the fork and branch.

Pros of 2: the GH PR number should be easily known by the runner configurer. 
Cons of 2: another runner configuration variable. Does not fully support running benchmarks outside of PR builds, or at least we'd need to handle those with more custom logic, including merge-commit and nightly runs. More server-side GitHub API requests.

(Somewhat unrelatedly: right now, `benchalerts` doesn't need to know anything about PR numbers. It just posts alerts on GitHub commits. Maybe someday in the future if we set up alerts to other services, it will need to figure out the PR number. But it could do that by hitting GitHub, not necessarily conbench.)

Reply:
Personally, I think "So it sounds like with either option, the client must scrape the repo name and commit sha from a local cloned repo" is even a bit too prescriptive even, the repo name is effectively static, so can be in an env var (or dot-conbench file), but sure needs to be set. As for the commit sha: that could come from a local git clone or it could come from asking the software-under-benchmark for it [like we do with pyarrow right now](https://github.com/voltrondata-labs/benchmarks/blob/b1aae88bcc1ca3f9163e2c229d7fafa453b39107/benchmarks/_benchmark.py#L38). IMO the second is a better option if we can do it that way since it guarantees the software being benchmarked is indeed at the revision one is expecting it to be — though there of course will be cases where software isn't configured to provide info like that.

I suspect that option 2 will serve us better in the future because we (as the backend) take on the onus of getting the details of that right at a point when we can relatively easily as opposed to requiring runners or adapters (or folks writing their own versions of both) always to have to get that exactly right.

Reply:
Got it, so what I'm hearing now (and I suspect what you've been saying the whole time) is that when it comes to supplying context, we should prefer more user-supplied configuration over scraping it from the machine setup, as long as the configuration is straightforward.

So with that design principle, here's my suggestion:
1. [optional] We remove the assumption that one conbench server can track multiple repos, and set `GITHUB_REPO_NAME="org/repo"` in the server config file.
2. The recommended way to provide GitHub info is to provide the commit sha and GitHub PR number when the runner POSTs a new run.
   - If we don't do (1), the runner must also provide the repo name in exactly that `org/repo` format.
   - If it's a merge-commit run, the runner provides `"github_pr_number": null` and we'll have to assume the commit sha is on the default branch.
   - We won't be able to handle runs on "special" non-default-branches (like commits to a `dev` branch) unless we add a third configuration flag. When we do want to add support for special branches, we'll probably have to do another overhaul PR with a migration, since `fork_point_sha` might not be as useful anymore.
3. We store the `fork:branch` string, and the fork point sha from the default branch, in the commit table.
4. If a runner is running benchmarks on a locally cloned repo, we could scrape all the necessary information from that repo and skip (2).

In this PR, I won't implement (1) yet. If we skip (4), it will be a breaking change for runners so we'll have to be extra careful about releasing.

Reply:
Yes, yes I like this. At one step further, it follows a principle that I haven't written down anywhere explicitly (I should) of trying to have the right thing filled in, but if someone really wants or knows what they are doing they can over-ride by sending the information for fields like these. Ideally no one will take the second path (because the first is easier + most likely to be right) but the option is there (which also allows us to over-ride it if we need to before we can get a backend fix in to enable something, for example).

A few comments about pieces: 

For (1) let's see if how helpful it is to make it so any one server has one and only one repo. One place we expected to use this (but haven't yet) is when benchmarking the arrow-rust implementation which is in `apache/arrow-rs`. We don't have to absolutely stick to this if it's super painful to keep sticking to it, but also if it doesn't help much to get rid of this feature, maybe we should keep it around? I'm very on the fence here

> If it's a merge-commit run, the runner provides "github_pr_number": null and we'll have to assume the commit sha is on the default branch.

This sounds right for (almost?) all cases right now.

> We won't be able to handle runs on "special" non-default-branches (like commits to a dev branch) unless we add a third configuration flag. When we do want to add support for special branches, we'll probably have to do another overhaul PR with a migration, since fork_point_sha might not be as useful anymore.

Would it be possible to set this up such that if someone supplies a branch element, we use that? This way someone _could_ start sending random branch names, though the comparison or history might not work with those until we overhaul? But in general we tell folks not to send branch name, and we get the behavior you describe here. If supporting both now is too much, that's totally fine, of course we can punt on being able to send an over-riding branch name for now and only accept commits to the default branch (which is effectively what the current system is restricted to anyway!)

> If a runner is running benchmarks on a locally cloned repo, we could scrape all the necessary information from that repo and skip (2).

Yeah, and this we should also mark as advanced mode + not recommended unless you are very very clear on the details and what it means.

Reply:
> 4\. If a runner is running benchmarks on a locally cloned repo, we could scrape all the necessary information from that repo and skip (2).

You can't scrape PR# afaik, you have to get it from CI. 

> * If it's a merge-commit run, the runner provides `"github_pr_number": null` and we'll have to assume the commit sha is on the default branch.

This is a little dangerous; that info is probably going to be coming through an env var in CI, and if it's set wrong or not set, it will be null even if it shouldn't be.

Reply:
Re: (1), it sounds like we'd really like to support multiple Arrow subprojects on the Arrow conbench (at least), so we can keep all the repo handling around. We'll just have to have runners provide the repo name when posting runs. (unless they're using the auto-scrape method)

>You can't scrape PR# afaik, you have to get it from CI.

Right sorry, I meant that we'd be able to auto-scrape repo, fork, branch, and commit from the local repo, so we don't need the PR number in that case. We're not storing PR number either way, just storing `fork:branch`.

>if someone supplies a branch element, we use that

Yeah this is probably the best way to do special branch handling, but let's hold off for now.

>`"github_pr_number": null` is a little dangerous

True as always. Luckily if a commit is supplied and `github_pr_number` is null, the server will (soon) be doing the default-branch backfill process, and we can validate the new commit is on the default branch during that process.

Reply:
> > "github_pr_number": null is a little dangerous

> True as always. Luckily if a commit is supplied and github_pr_number is null, the server will (soon) be doing the default-branch backfill process, and we can validate the new commit is on the default branch during that process.

Yeah, and having a nice message there that says "Hey, you sent me a commit but that commit is nowhere on the default branch, something is wrong, are you trying to send a PR branch? If so..." will hopefully reduce confusion and some of the magic that we have now that makes it even more confusing...

Reply:
Okay, this is implemented. If you all have more comments on the implementation let's start a new thread :)